rewrites http client and server to use libh2o

joemfb opened the pull request:
This is a drop-in, compatible replacement of the custom http server with the mature implementation of H2O, notably adding TLS termination and the potential for straightforward HTTP 2 and Websocket support.

This PR is *NOT* ready to be merged until a dependency management decision is made; to build this branch for now, download [the 2.2.3 release of H2O](https://github.com/h2o/h2o/releases/tag/v2.2.3) and extract the archive to `outside/h2o-2.2.3/`.

That dependency could be bundled, but I've deferred in light of #934. libh2o packages appear to be relatively available:

- https://packages.ubuntu.com/bionic/libh2o0.13
- https://packages.debian.org/sid/libh2o0.13
- https://bintray.com/tatsushid/h2o-rpm/libh2o
- https://github.com/Homebrew/homebrew-core/blob/master/Formula/h2o.rb

I've tested this PR extensively with AddressSanitizer enabled, and everything appears to be in working order.

It's worth noting that the terminal event spinner now has different behavior on incoming http requests. Since this implementation is not reading directly from the socket, I'm not calling `u3_lo_open()` and `u3_lo_shut()` in the usual manner. Instead, pending http events are processed on every turn of the event loop due to the `uv_check_t` handler in `unix.c` (see `_unix_ef_sync`). As a result, the terminal spinner appears briefly for each individual request in a page load, as opposed to (apparently) coalescing the loop blocking for that group of requests. I think this topic merits a larger discussion (in which #910 and cc-release both have a role), but I'm not quite prepared to lead said discussion.

Additional work will be needed to implement a more robust approach to certificate management. But I'm opening this PR now because a) it's been in progress for awhile, b) it's ~done as a replacement for existing functionality, and c) the remaining determinations intersect with other open PRs.

joemfb commented:
For a more readable diff (pre-dating more aggressive refactoring): https://github.com/urbit/urbit/compare/master...joemfb:159d586

cgyarvin commented:
I will look at this week.  I don't see why it can't be pulled -- especially
considering the relatively available and unbundled package...

On Tue, Mar 6, 2018 at 2:48 PM, Joe Bryan <notifications@github.com> wrote:

> For a more readable diff (pre-dating more aggressive refactoring):
> master...joemfb:159d586
> <https://github.com/urbit/urbit/compare/master...joemfb:159d586>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/urbit/pull/942#issuecomment-370957111>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALyAfPuJsP6yUQWgode6OxWv0bBzWJCks5tbxIlgaJpZM4SfkOj>
> .
>

mikolajpp commented:
Adding this to meson, and bundling it in case it is missing is just one line.

Will do so in the evening. For the meson exercise, consider me "the meson guy", until it makes into master. 

Afterwards we can just forget and it should work

mikolajpp commented:
Okay, it would be good for either of PRs to first make it to the master to not pollinate commits...

I think it would be best to merge meson build first, and then integrate this, since in meson build we change a structure a bit, and bundling of source in urbit repo is no longer needed. If anything, we put missing deps as git submodules which are pulled on demand

joemfb commented:
@mikolajpp, meson's approach to specifying a dependency with a bundled fallback looks really nice. I'm inclined to wait for your PR to be merged based on that alone.

I should note that after this PR is merged, `cttp.c` will be the only code using `http_parser`. I'm looking at how viable an H2O-based replacement client would be.

joemfb commented:
I've replaced the client as well, and refactored the server in little in light of those changes.

I've tested the client in :dojo with AddressSanitizer, and everything seems to be working correctly. But I still need to write an app so I can test more thoroughly, check the cancellation, etc.

This should fix #838.

eglaysher commented:
OK, the meson build is now merged. As far as I can tell, all this patch needs before merging is updating the build system / adding an h2o submodule dependency.

joemfb commented:
@eglaysher, do you have a preference between rebasing this on top of the build changes vs merging master into this branch and then patching the build? (The first option would probably make it easier to bisect and build the interim commits ...)

eglaysher commented:
I have a slight preference for merges. You already wrote these patches. But do whatever is easiest for you.

joemfb commented:
I've backed out the meson changes from the PR. They're still available at https://github.com/joemfb/urbit/tree/libh2o-meson.

joemfb commented:
I've backed out the meson changes from the PR. They're still available at https://github.com/joemfb/urbit/tree/libh2o-meson.

Fang- commented:
@joemfb, you mention this fixing #838, but I can still reproduce it, both using the repro steps in that issue, and by running my own WIP code that does HTTP requests. This was tested using a binary built from your meson branch, the branch this PR points to gave me an out of memory error.

joemfb commented:
@Fang-, are you crashing and dumping core, or just getting failures returned as 504s?

Fang- commented:
Proper coredumps.

```
Assertion failed: (handle->type == UV_TCP || handle->type == UV_TTY || handle->type == UV_NAMED_PIPE), function uv___stream_fd, file src/unix/stream.c, line 1632.
fish: '~/softsrc/urbit/build/urbit -FI…' terminated by signal SIGABRT (Abort)
```

Fang- commented:
```
(lldb) bt
* thread #1, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff9ef2ed42 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff9f01c457 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff9ee94420 libsystem_c.dylib`abort + 129
    frame #3: 0x00007fff9ee5b893 libsystem_c.dylib`__assert_rtn + 320
    frame #4: 0x0000000109bc4dd0 libuv.1.dylib`uv___stream_fd + 84
    frame #5: 0x0000000109bc3d42 libuv.1.dylib`uv__stream_io + 1364
    frame #6: 0x0000000109bcb426 libuv.1.dylib`uv__io_poll + 1744
    frame #7: 0x0000000109bbc47a libuv.1.dylib`uv_run + 333
    frame #8: 0x0000000109791315 urbit`main(argc=4, argv=0x00007fff56479a40) at main.c:624 [opt]
    frame #9: 0x00007fff9ee00235 libdyld.dylib`start + 1
```

joemfb commented:
This is all up-to-date once again, but it will require some work to get the libh2o fallback subproject working in all scenarios. On OS X, it has the same trouble finding openssl (and will either need the PKG_CONFIG_PATH or explicit fallback paths), and here it can't find libuv, since that's now a sibling subproject and not an installed system library.

@mikolajpp, do you know any special tricks for sharing subproject dependencies across subprojects in meson?

joemfb commented:
For anyone using homebrew on OS X, the simplest way to try this branch is to first `brew install h2o --with-libuv`

joemfb commented:
If we just replicate the subproject fallback from the root meson.build in
the libh2o subproject, we’ll end up checking out and building libuv twice.
Similarly, we’d have to duplicate the OpenSSL workaround for macOS. It’d be
nicer to only declare those once, preferably at the top level or as
siblings, but I can’t see how to do that.

The next best thing would be to declare the libuv and OpenSSL dependencies
in the libh2o meson.build, and just reference them at the top level. That’s
not as nice, and a little strange, but it should work. I’ll try it in the
morning.

On Mon, Apr 9, 2018 at 11:21 PM mikolajpp <notifications@github.com> wrote:

> @joemfb <https://github.com/joemfb> - no tricks - just the standard way.
>
> For subprojects to work, see how it works in meson.build - there is an
> extended syntax dependency with 'fallback' field. you have to use this
> syntax in the meson.build for h2o itself.
>
> If you won't manage to get it work, i will have some time this week.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/urbit/pull/942#issuecomment-379987163>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAq3YHQEUXMXk-q6mlNBp2uksPqmXXTdks5tnE9PgaJpZM4SfkOj>
> .
>
-- 
Sent from Gmail Mobile

mikolajpp commented:
@joemfb 

Why would it? Meson is smart enough to figure dependency tree and reuse it. 
I just did it and it compiles just fine. Here is what you have to put in libh2o meson.build

```
libuv_dep = dependency('libuv', version: '>=1.8.0', fallback:['libuv', 'libuv_dep'])
pthread_dep = meson.get_compiler('c').find_library('pthread')

lib = library('libh2o', sources: lib_src, version: version,
include_directories: inc,
c_args: ['-Wall', '-Wno-unused-value', '-Wno-unused-function', '-O2'],
dependencies: [libuv_dep, zlib_dep, openssl_dep, pthread_dep],
install: true)
```

The same goes for openssl.

mikolajpp commented:
@joemfb 
I will be out for this week, so if another meson conundrum happens, contact me the Urbit way :P

mikolajpp commented:
In general requiring dependency the second time from a subproject will just reference back to already detected version. Under no circumstances it could trigger a second build. When you declare dependency in meson nothing actually happens - it just goes to the work list. 

There is only one subproject for `libuv` - hence it will get build exactly once.

mikolajpp commented:
So the proper way to handle all this is just declare dependencies where we hit them the first time.
I have no idea what other parts of urbit require openssl/libuv. 

If it is only libh2o => Then these deps should get removed from root meson.build and put under libh2o meson.build. 

If something else requires these deps, they should be put the first time we require them - if something is required by urbit src, then we should also declare them in root meson.build, and so on.

ohAitch commented:
- **libuv** [is used](https://github.com/urbit/urbit/search?q=uv) throughout `vere/`, for all console/UDP/timeouts/etc. event loop interaction, whether or not we're using it for HTTP in particular.

- **openssl** looks like it's only referenced in the http client `cttp`, so as long as `libh2o` handles all ssl inbound and outbound we shouldn't need it as a direct dependency. (It's also in vere `-R`, but as a not-direct-dependency can presumably be pulled out. Though it looks like this PR adds some `random` initialization in `main`, which talks to `openssl` directly without going through an `h2o` API?)

joemfb commented:
`http.c`, `cttp.c`, and libh2o all directly depend on openssl. The client and server have to initialize the ssl context explicitly before passing it into the library.

The CPRNG initialization that's now in `main.c` used to be in `cttp.c`, but I moved it since both the client and server are now using openssl. I don't think the seeding is strictly necessary, as OpenSSL automatically uses `/dev/urandom` where available, but 1) it can't hurt and 2) it was already there, and I figured we should discuss before removing it ... ;)

Fang- commented:
Curious, what's the reason we're hosting a fork/clone of h2o ourselves rather than pointing to its original repo? Submodules get pinned at specific commits, if I'm not mistaken, so we could still update at our own pace.

joemfb commented:
Because we're adding meson build configuration for each of those repositories. I'm not a big fan of it myself, and maybe we could look into the dependency wrap and wraps-db features of meson, so we don't have to have all these forks ...

keatondunsford commented:
@joemfb Does this PR require an update to the Install doc? (e.g. installing the `h2o` dependency)

keatondunsford commented:
Oh, it's just a submodule.

joemfb commented:
You can install the h2o library as a dependency, but it's not required -- the build falls back to the subproject.

keatondunsford commented:
Cool.

keatondunsford commented:
Clarification: whereas before this merge, the start prompt would print this:

```
http: live ("secure", public) on 8443
```

is it normal that I'm now getting this?:

```
http: secure server error: no tls config
```

ohAitch commented:
yes, though that should probably be rephrased to be less alarming, e.g. "http: secure server not starting, .urb/tls not found"

joemfb commented:
I've opened #965 to clean up that message. Let me know if you any other printfs should changed too (for instance, the earlier "https: failed to load certificate".

keatondunsford commented:
Error message improved with #965.
